### PR TITLE
Add code to generate SPLASH golden datasets

### LINKED ISSUE #184: Add code used to generate 'golden' datasets using the original SPLASH implentation into `pyrealm_build_data/splash`

### PR DESCRIPTION
**Pull request by davidorme**
# Description

This PR fixes #184. The changes:

* Add a local copy of the original SPLASH code into `pyrealm_build_data/splash/splash_py_version`. This has been lightly modified to allow it to be imported as a package, and some code formatting has happened but this code is _deliberately_ excluded from `mypy` and `flake8` checking. 
* Python scripts have been added to `pyrealm_build_data/splash` that generate or format the test input datasets and then feed the data into the original SPLASH code to generate the golden datasets.
* A `PROVENANCE.md` file documents the source and modification of the original code.
* A `README.md` file describes the commands used to generate the golden datasets.

ETA: Although there are a lot of files changed, many of them are data files and the code files in `splash_py_version` **should not be reviewed** as they are intentionally a clone (with very minor edits for paths and formatting style) of the reference implementation.

Still to do:

- [x] Track down issue that is causing the current golden outputs to differ slightly from those generated by the added files.
- [x] Tidy up - probably splitting files in `pyrealm_build_data/splash` into  `scripts` and the existing `data` folder. 

## Type of change

- [x] New feature (non-breaking change which adds functionality)
- [ ] Optimization (back-end change that speeds up the code)
- [ ] Bug fix (non-breaking change which fixes an issue)

## Key checklist

- [x] Make sure you've run the `pre-commit` checks: `$ pre-commit run -a`
- [x] All tests pass: `$ poetry run pytest`

## Further checks

- [x] Code is commented, particularly in hard-to-understand areas
- [x] Tests added that prove fix is effective or that feature works

### CODE REVIEW

**Comment by davidorme** [comment]
Well that was no fun.

I was unable to duplicate the original test datasets when re-running the dataset creation. Nothing major but small float errors. Turns out I was passing the Julian day of the year into the old SPLASH `run_one_day` function as a single value `xarray.DataArray` rather than as an integer. That turned out to have a massive performance hit (960 seconds -> 26 seconds) but also seems to have done something truly odd with the calculations resulting in those floating point issues. It’s an int64 DataArray, so it isn’t being treated as a float. No idea what is going on here, but very puzzling, and wasn't happening when I created the datasets originally.

**Comment by davidorme** [comment]
> Could an alternative way of storing the data (figshare or Zenodo, for example) make sense?

Definitely could. I've tried to keep them reasonably small - and they are deliberately not in the main `pyrealm` package so that they don't bloat it. They get used in the documentation as annotated examples for usage, but I guess we could have a roll-your-own LFS style thing that fetches a Zenodo deposit? I'm inclined to leave it for now, but it is a good point.

**Comment by MarionBWeinzierl** [comment]
> > Could an alternative way of storing the data (figshare or Zenodo, for example) make sense?
> 
> Definitely could. I've tried to keep them reasonably small - and they are deliberately not in the main `pyrealm` package so that they don't bloat it. They get used in the documentation as annotated examples for usage, but I guess we could have a roll-your-own LFS style thing that fetches a Zenodo deposit? I'm inclined to leave it for now, but it is a good point.

I created an issue (#189) for this and put it into the icebox for now.

File: pyrealm_build_data/splash/splash_py_version/PROVENANCE.md [hunk 1]
```diff
@@ -0,0 +1,18 @@
+# Provenance of the splash_py_version directory
+
+This directory contains the original Python implementation of the SPLASH model,
+developed by Tyler Davis, hosted at [https://bitbucket.org/labprentice/splash]. The
+files here are the contents of the path: `releases/v1.0/py_version` from the `master`
+branch of the repo at commit `52d9454b566d`.
+
+Changes:
+
+* This note and the `__init__.py` file to allow the module to be imported are the  only
```
Comment (surbhigoel77) [inline pyrealm_build_data/splash/splash_py_version/PROVENANCE.md:10]: Maybe rephrase this sentence?

File: pyrealm_build_data/splash/splash_py_version/splash_data.py [hunk 2]
```diff
@@ -0,0 +1,834 @@
...
+        Name:     SPLASH_DATA.density_h2o
+        Input:    - float, air temperature (tc), degrees C
+                  - float, atmospheric pressure (p), Pa
+        Output:   float, density of water, kg/m^3
+        Features: Calculates density of water at a given temperature and
+                  pressure
+        Ref:      Chen et al. (1977)
+        """
+        # Calculate density at 1 atm:
+        po = (
```
Comment (surbhigoel77) [inline pyrealm_build_data/splash/splash_py_version/splash_data.py:264]: Number of multiplications can be reduced using `Horner's method`. Basically taking out common `tc` at every step.

po = (
    0.99983952
    + tc * (
        6.788260e-5
        + tc * (
            -9.08659e-6
            + tc * (
                1.022130e-7
                + tc * (
                    -1.35439e-9
                    + tc * (
                        1.471150e-11
                        + tc * (
                            -1.11663e-13
                            + tc * (
                                5.044070e-16
                                - tc * (
                                    1.00659e-18 * tc
                                )
                            )
                        )
                    )
                )
            )
        )
    )
)

File: pyrealm_build_data/splash/splash_py_version/splash_data.py [hunk 3]
```diff
@@ -0,0 +1,834 @@
...
+        Name:     SPLASH_DATA.set_tair_source
+        Input:    str, source preference (n)
+        Features: Define the source preference for air temperature (e.g., 'cru'
+                  or 'watch')
+        """
+        logging.info("setting temperature preference to '%s'", n)
+        self.data_src["tair"] = n
+
+    def add_one_month(self, dt0):
+        """
```
Comment (surbhigoel77) [inline pyrealm_build_data/splash/splash_py_version/splash_data.py:239]: We can mention the necessity to add/sub a month to the datetime.

Comment (davidorme) [reply pyrealm_build_data/splash/splash_py_version/splash_data.py:239]: @surbhigoel77 - Don't review the code files in `splash_py_version` - there's all sorts of oddities in the code, but its the benchmark implementation so I'm leaving them as is.

File: pyrealm_build_data/splash/splash_run_time_series_parallel.py [hunk 4]
```diff
@@ -0,0 +1,258 @@
...
+    """Run a SPLASH time series for gridded data."""
+
+    start = datetime.now()
+    n_real = 0
+
+    # Import the test data file
+    input_data = xarray.load_dataset(input_file)
+
+    # Create storage for the results: evap and core solar
+    # Create storage for the results: evap and core solar
```
Comment (tztsai) [inline pyrealm_build_data/splash/splash_run_time_series_parallel.py:155]: Duplicate line

File: pyrealm_build_data/splash/splash_run_time_series_parallel.py [hunk 5]
```diff
@@ -0,0 +1,258 @@
...
+            cell_elev = input_data["elev"][lat_idx, lon_idx]
+            coords.append(
+                (
+                    lat_idx,
+                    float(lat.data),
+                    lon_idx,
+                    float(lon.data),
+                    float(cell_elev.data),
+                )
+            )
```
Comment (tztsai) [inline pyrealm_build_data/splash/splash_run_time_series_parallel.py:186]: Maybe a list comprehension `coords = [(float(lat.data), float(lon.data), float(cell_elev.data)) for lat in input_data.lat for lon in input_data.lon]` would be more concise.

File: pyrealm_build_data/splash/splash_run_time_series_parallel.py [hunk 6]
```diff
@@ -0,0 +1,258 @@
...
+            "econ": econ,
+            "cond": cond,
+            "eet_d": eet,
+            "pet_d": pet,
+            "aet_d": aet,
+            "wn": wn,
+            "ro": ro,
+            "wn_spun_up": wn_spun_up,
+        }
+    )
```
Comment (tztsai) [inline pyrealm_build_data/splash/splash_run_time_series_parallel.py:238]: I feel this could be more concise by using `df = pd.DataFrame(results)` followed by `out = df.to_xarray()`.

**Review by tztsai — COMMENTED** [review]
Hi @davidorme please check out the reciews.

File: pyrealm_build_data/splash/splash_py_version/splash_data.py (hunk 2 above)
Comment (davidorme) [reply pyrealm_build_data/splash/splash_py_version/splash_data.py:264]: It absolutely can - but this is one of the 'original' implementation files where I want to leave the process untouched.

**Comment by davidorme** [comment]
> Why is `splash_py_version` added to clone the functionality of `pyrealm.splash`?

@tztsai That directory contains a previous implementation that had all sorts of implementation and usability issues. This PR is all about using the previous version to generate some regression test inputs to validate I have re-implemented the approach successfully.

File: pyrealm_build_data/splash/splash_run_time_series_parallel.py (hunk 5 above)
Comment (davidorme) [reply pyrealm_build_data/splash/splash_run_time_series_parallel.py:186]: That's leaving out the indices and the `elev` extraction. Its on the borderline between list comprehension and loops being simpler to comprehend, but I think this _is_ a bit cleaner:

```
coords = [
    (
        lat_idx,
        float(lat.data),
        lon_idx,
        float(lon.data),
        input_data["elev"].data[lat_idx, lon_idx],
    )
    for lat_idx, lat in enumerate(input_data.lat)
    for lon_idx, lon in enumerate(input_data.lon)
]
```

Comment (tztsai) [reply pyrealm_build_data/splash/splash_run_time_series_parallel.py:186]: Thanks for the response! However, if `results` is directly converted to a `pd.DataFrame` as shown in the next comment, `lat_idx` and `lon_idx` would not be necessary.

Comment (davidorme) [reply pyrealm_build_data/splash/splash_run_time_series_parallel.py]: Except that it is need to isolate `cell_elev` from `input_data['elev'] and that is not included in your list comprehension.

Comment (davidorme) [reply pyrealm_build_data/splash/splash_run_time_series_parallel.py]: I guess the other thing is that this is a rarely used setup/testing script, so doesn't need finessing to quite the same degree as main package code.

**Comment by davidorme** [comment]
@surbhigoel77 and @tztsai - Can someone give an approving review so I can merge 😄

I've made a couple of minor additions to make it easier to generate benchmarks and access the reference implementation from other code.

**Review by surbhigoel77 — APPROVED** [review]
